docs/src/main/sphinx/sql/match-recognize.rst

```diff
@@ -0,0 +1,662 @@
+
+After match skip
+----------------
+
+The ``AFTER MATCH SKIP`` clause specifies where pattern matching resumes after
+a non-empty match is found.
+
+The default option is::
+
+    AFTER MATCH SKIP PAST LAST ROW
+
+With this option, pattern matching starts from the row after the last row of
+the match. Overlapping matches are not detected.
+
+With the following option, pattern matching starts from the second row of the
+match::
+
+    AFTER MATCH SKIP TO NEXT ROW
+
+In the example, if a V-shape is detected, further overlapping matches are
+found, starting from consecutive rows on the descending slope of the "V".
+Skipping to the next row is the default behavior after detecting an empty match
+or unmatched row.
+
+The following ``AFTER MATCH SKIP`` options allow to resume pattern matching
+based on the components of the pattern. Pattern matching starts from the last
+(default) or first row matched to a certain row pattern variable. It can be
+either a primary pattern variable (they are explained in
+:ref:`row_pattern_syntax`) or a
+:ref:`union variable<row_pattern_union_variables>`::
+
+    AFTER MATCH SKIP TO [ FIRST | LAST ] pattern_variable
+
+It is forbidden to skip to the first row of the current match, because it
+results in an infinite loop. For example specifying ``AFTER MATCH SKIP TO A``
+fails, because ``A`` is the first element of the pattern, and jumping back to
+it creates an infinite loop. Similarly, skipping to a pattern variable which is
+not present in the match causes failure.
+
+All other options than the default ``AFTER MATCH SKIP PAST LAST ROW`` allow
+detection of overlapping matches. The combination of ``ALL ROWS PER MATCH WITH
+UNMATCHED ROWS`` with ``AFTER MATCH SKIP PAST LAST ROW`` is the only
+configuration that guarantees exactly one output row for each input row.
+
+.. _row_pattern_syntax:
+
+Row pattern syntax
+------------------
+
+Row pattern is a form of a regular expression with some syntactical extensions
+specific to row pattern recognition. It is specified in the ``PATTERN``
+clause::
+
+    PATTERN ( row_pattern )
+
+The basic element of row pattern is a primary pattern variable. Like pattern
+matching in character strings searches for characters, pattern matching in row
+sequences searches for rows which can be "labeled" with certain primary pattern
+variables. A primary pattern variable has a form of an identifier and is
+:ref:`defined<row_pattern_variable_definitions>` by a boolean condition. This
+condition determines whether a particular input row can be mapped to this
+variable and take part in the match.
+
+In the example ``PATTERN (A B+ C+ D+)``, there are four primary pattern
+variables: ``A``, ``B``, ``C``, and ``D``.
+
+Row pattern syntax includes the following usage:
+
+concatenation
+^^^^^^^^^^^^^
+
+It is a sequence of components without operators between them:
+
+.. code-block:: text
+
+    A B+ C+ D+
+
+alternation
+^^^^^^^^^^^
+
+The order of alternatives defines the order of preference:
```

Comment:
This is confusing. We should describe what is an alternation and then give this additional detail on their behavior in case of ambiguous matches.

Reply:
I fixed the descriptions of alternation, concatenation and permutation.